zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2788,40 +2803,124 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
+        if (outputStat != null) {
+            DataTree.copyStat(response.getStat(), outputStat);
         }
+        updateNextPage(outputNextPage, response.getNextPageCzxid(), response.getNextPageCzxidOffset());
+
         return response.getChildren();
     }
 
+    /**
+     * Returns a list of all the children given the path.
+     * <p>
+     * The difference between this API and {@link #getChildren(String, boolean)} is:
+     * when there are lots of children and the network buffer exceeds {@code jute.maxbuffer},
+     * this API will fetch the children using pagination and be able to return all children;
+     * while {@link #getChildren(String, boolean)} will fail.
+     * <p>
+     * The final list of children returned is NOT strongly consistent with the server's data:
+     * the list might contain some deleted children if some children are deleted before the last page is fetched.
+     * <p>
+     * If the watch is true and the call is successful (no exception is thrown),
+     # a watch will be left on the node with the given path. The watch will be
+     * triggered by a successful operation that deletes the node of the given
+     * path or creates/deletes a child under the node.
+     *
+     * @param path the path of the parent node
+     * @param watch whether or not leave a watch on the given node
+     * @return a unordered list of all children of the given path
+     * @throws KeeperException if the server signals an error with a non-zero error code.
+     * @throws InterruptedException if the server transaction is interrupted.
+     */
+    public List<String> getAllChildrenPaginated(String path, boolean watch)
+            throws KeeperException, InterruptedException {
+        PaginationNextPage nextPage = new PaginationNextPage();
+        Watcher watcher = watch ? watchManager.defaultWatcher : null;
+        List<String> childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, 0L, 0, null, nextPage);
+        if (nextPage.getMinCzxid() == ZooDefs.GetChildrenPaginated.lastPageMinCzxid) {
+            return childrenPaginated;
+        }
+
+        Set<String> childrenSet = new HashSet<>(childrenPaginated);
+        getChildrenRemainingPages(path, watcher, nextPage, childrenSet);
+
+        return new ArrayList<>(childrenSet);
+    }
+
+    private void getChildrenRemainingPages(String path,
+                                           Watcher watcher,
+                                           PaginationNextPage nextPage,
+                                           Set<String> childrenSet) throws KeeperException, InterruptedException {
+        int retryCount = 0;
+        do {
+            List<String> childrenPaginated;
+            long minCzxid = nextPage.getMinCzxid();
+            if (nextPage.getMinCzxidOffset() == 0) {
+                childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, minCzxid,
+                        nextPage.getMinCzxidOffset(), null, nextPage);
+            } else {
```

Comment:
Let's reorganize the logic. It looks wired. It should be:

 -> nextPage.getMinCzxidOffset() == 0
        Y -> return it.
        N ->  handle it with 3 retries.

So I image the code is in following structure. That could be more clear with different conditions.

if (nextPage.getMinCzxidOffset() == 0) {
               childrenSet.addAll(getChildren(path, watcher, Integer.MAX_VALUE, minCzxid,
                        nextPage.getMinCzxidOffset(), null, nextPage));
               return;
}

do {
    xxxxxxx
} while (xxx);
return childrenPaginated;

Reply:
@dasahcc Thanks for the suggestion. That'd be local retries - each czxid will have 3 retries. I indeed thought about this and tried it previously, but it would make the code more complicated because it requires more variables and logic to handle the retries for the same local czxid.

And now this logic is a global retry - globally the getChildren operation can only retry 5 times in total. Considering we are not using transaction to create children and we are not writing children with 10 MB jute maxbuffer and change it back to 1 MB, I would say, global 5 retries in total is good enough. It makes the code less and cleaner.

Reply:
For single page getChildren, why we need to do single retry? If it is failed to read, we should fail it because of different other reasons, right?

Reply:
@dasahcc You might be confused. This part is for getting all children with potentially multiple pages, not a single page getChildren. You are right, single page getChildren does not have any retry.